### PR TITLE
tieredReplicants can be null

### PR DESCRIPTION
**Pull request by pjain1**
- Setting tieredReplicants to be an empty Map if it is null so that getNumReplicants method does not throws NPE.
- Added serde test for when it is null

### CODE REVIEW

File: server/src/main/java/io/druid/server/coordinator/rules/IntervalLoadRule.java [hunk 1]
```diff
@@ -41,9 +42,9 @@ public IntervalLoadRule(
       @JsonProperty("tieredReplicants") Map<String, Integer> tieredReplicants
   )
   {
-    validateTieredReplicants(tieredReplicants);
+    this.tieredReplicants = tieredReplicants == null ? Maps.<String, Integer>newHashMap() : tieredReplicants;
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/coordinator/rules/IntervalLoadRule.java]: Suggest `ImmutableMap.<String, Integer>of()`

File: server/src/test/java/io/druid/server/coordinator/rules/ForeverLoadRuleTest.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    ForeverLoadRule rule = new ForeverLoadRule(
+        null
+    );
+
+    ObjectMapper jsonMapper = new DefaultObjectMapper();
+    Rule reread = jsonMapper.readValue(jsonMapper.writeValueAsString(rule), Rule.class);
+
+    Assert.assertEquals(rule.getTieredReplicants(), ((ForeverLoadRule)reread).getTieredReplicants());
+  }
+}
```
Comment (drcrallen) [inline server/src/test/java/io/druid/server/coordinator/rules/ForeverLoadRuleTest.java]: missing newline

File: server/src/main/java/io/druid/server/coordinator/rules/IntervalLoadRule.java (hunk 1 above)
Comment (pjain1) [inline server/src/main/java/io/druid/server/coordinator/rules/IntervalLoadRule.java]: sure

File: server/src/test/java/io/druid/server/coordinator/rules/ForeverLoadRuleTest.java (hunk 2 above)
Comment (pjain1) [inline server/src/test/java/io/druid/server/coordinator/rules/ForeverLoadRuleTest.java]: ok

**Comment by drcrallen** [comment]
Thanks for the fix!

Can you clarify the following by way of unit tests:

1. If a `tieredReplicants ` is NULL in json, should the resulting property be null or empty-map?
1. If a `tieredReplicants` is NULL, should reflecting json result in the same original json?

In the prior behavior it did result in the same json. If I'm reading this correctly this PR changes such behavior.

There is precedence for such behavior in numerous other POJOs around Druid, but it would be nice to get the weigh-in of someone more familiar with the Replicant Tiering code to comment on if this change will be an issue.

**Comment by pjain1** [comment]
@drcrallen Yes it changes the behavior. `tieredReplicants` can be set to `null` instead of an empty Map however in that case wouldn't `getNumReplicants(String tier)` method and `getTieredReplicants()` in `run` method of `LoadRule` throw NPE?

**Comment by drcrallen** [comment]
@pjain1 : Yes, You are correct. The case we encountered was for datasources which were not present, but still had legacy database entries. As such we did not encounter the NPE.

I looked through and could not find an obvious place where the rule would cause an issue if it did not have a value here and suddenly does, and it makes it more of a passive upgrade method in case old dormant systems kick back online. As such this should be reasonable. But can you please still add unit tests for checking results of what's in `tieredReplicants` to make sure it is a map if a null is passed in (instead of just checking for equality between two higher level objects)?

**Comment by pjain1** [comment]
@drcrallen sure I can do that.

**Comment by pjain1** [comment]
@drcrallen I've added the tests you mentioned. Please verify.

File: server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java [hunk 3]
```diff
@@ -243,6 +243,8 @@ public void execute()
 
   protected void validateTieredReplicants(Map<String, Integer> tieredReplicants){
     for (Map.Entry<String, Integer> entry: tieredReplicants.entrySet()) {
+      if (entry.getValue() == null)
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java]: This was added but not tested. Can you add at least one test that gets to this codepath?

File: server/src/main/java/io/druid/server/coordinator/rules/IntervalLoadRule.java [hunk 4]
```diff
@@ -41,9 +42,9 @@ public IntervalLoadRule(
       @JsonProperty("tieredReplicants") Map<String, Integer> tieredReplicants
   )
   {
-    validateTieredReplicants(tieredReplicants);
+    this.tieredReplicants = tieredReplicants == null ? ImmutableMap.<String, Integer>of() : tieredReplicants;
```
Comment (xvrl) [inline server/src/main/java/io/druid/server/coordinator/rules/IntervalLoadRule.java]: I wonder if the proper default for tiers should be `DruidServer.DEFAULT_TIER` and `DruidServer.DEFAULT_NUM_REPLICANTS`. That would be more in line with our existing defaults.

Comment (xvrl) [inline server/src/main/java/io/druid/server/coordinator/rules/IntervalLoadRule.java]: More specifically ImmutableMap.of(DruidServer.DEFAULT_TIER, DruidServer.DEFAULT_NUM_REPLICANTS)

Comment (drcrallen) [inline server/src/main/java/io/druid/server/coordinator/rules/IntervalLoadRule.java]: That's a good call

Comment (xvrl) [inline server/src/main/java/io/druid/server/coordinator/rules/IntervalLoadRule.java]: I also think we should not allow empty replicants. It wouldn't make sense to have a load rule without specifying a place to put the data. `null` should mean "use the default" and empty should not be allowed.

The only caveat would be if someone created rules with empty tieredReplicant map in the past, that would cause the coordinator to fail miserably, as it did for us with null values.

File: server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java (hunk 3 above)
Comment (pjain1) [inline server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java]: Done

File: server/src/main/java/io/druid/server/coordinator/rules/IntervalLoadRule.java (hunk 4 above)
Comment (pjain1) [inline server/src/main/java/io/druid/server/coordinator/rules/IntervalLoadRule.java]: I have made changes to convert the null to use default values and not allow empty ones. Can you please check if not allowing empty values cause any problem with the legacy rules in your system.

File: server/src/test/java/io/druid/server/coordinator/rules/ForeverLoadRuleTest.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
+package io.druid.server.coordinator.rules;
```
Comment (fjy) [inline server/src/test/java/io/druid/server/coordinator/rules/ForeverLoadRuleTest.java]: license header

Comment (pjain1) [inline server/src/test/java/io/druid/server/coordinator/rules/ForeverLoadRuleTest.java]: done

**Comment by drcrallen** [comment]
LGTM waiting for travis CI

**Comment by pjain1** [comment]
Squashed the commits.

**Comment by fjy** [comment]
:+1:

**Comment by xvrl** [comment]
LGTM
